Add ability to import JXL images

xiota opened the pull request:
This allows RawTherapee to import JXL images.  Addresses [the import portion] of #6273.

Thanatomanic commented:
@xiota Thanks for this! 

> Addresses half of #6273.

Which half? And can you elaborate a bit about what is still missing?

xiota commented:
#6273 requests import *and* export.  This is for import only.  Hence half.

Floessie commented:
@xiota Thanks for your contribution! :+1: There seem to be some (error) conditions under which `ImageIO::loadJxl()` leaks allocated memory in `buffer`. Could you double-check?

xiota commented:
@Floessie Thanks for letting me know.  I added a couple lines to free the buffer.

Thanatomanic commented:
@xiota Just out of curiosity, do you have a particular reason to force-push the last few commits instead of committing regularly with descriptions? We can squash-merge things in the end so the master commit log won't be polluted.

xiota commented:
@Thanatomanic Was just to keep the commit log clean.  Didn't know you would squash merge.

xiota commented:
This was working yesterday, but it isn't working after I recompiled today without making any changes.

xiota commented:
It's working again.  I was trying to make compiling with libjxl optional, but something was being cached, so it looked like it was working when it wasn't.  Someone else will have to make the changes needed to make cmake detect and load libjxl as needed.

Lawrence37 commented on rtgui/thumbnail.cc:
```diff
@@ -136,50 +135,37 @@ void Thumbnail::_generateThumbnailImage ()
+    if ( tpp == nullptr ) {
+        quick = false;
+        tpp = rtengine::Thumbnail::loadFromRaw (fname, ri, sensorType, tw, th, 1, pparams->wb.equal, TRUE);
+    }
```
Doesn't this lead to unnecessary function calls, compared to the old code?

xiota replied:
It will call functions to check if the file is a raw before trying to open non-raw file types.  But...

* It reduces nesting by removing the if-else-if-else-if-else structure.
* It allows formats to be added to imageio without having to return to this particular section to add file extensions.
* This is a raw processor, so prioritizing raw files seems reasonable.
* The raw-processing functions should exit quickly enough for non-raw files.

Lawrence37 replied:
Ok, if the raw functions exit quickly for non-raw files, it should be fine. Thanks for clarifying.

Lawrence37 replied:
@xiota FYI this may be causing some jpg files to open incorrectly because they are treated as raw files. See https://discuss.pixls.us/t/rt-problem-nikon-z8-w-z50mm-1-8s-lens/43933

xiota replied:
Thanks.  Will look at more.

xiota replied:
I have a potential fix.  Do you want it in a separate PR?  Would it be okay to add to #7096, along with a couple other misc changes, like a workflow update that needs to be in dev to make testing libjxl 0.7.x easier.

Lawrence37 commented on rtdata/options/options.lin:
```diff
@@ -12,8 +12,8 @@ MultiUser=true
 # Image filename extensions to be looked for, and their corresponding search state (0/1 -> skip/include)
-ParseExtensions=3fr;arw;arq;cr2;cr3;crf;crw;dcr;dng;fff;iiq;jpg;jpeg;kdc;mef;mos;mrw;nef;nrw;orf;ori;pef;png;raf;raw;rw2;rwl;rwz;sr2;srf;srw;tif;tiff;x3f;
-ParseExtensionsEnabled=1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;0;1;0;1;1;1;1;1;1;1;1;1;1;1;
+ParseExtensions=3fr;arw;arq;cr2;cr3;crf;crw;dcr;dng;fff;iiq;jpg;jpeg;kdc;mef;mos;mrw;nef;nrw;orf;ori;pef;png;raf;raw;rw2;rwl;rwz;sr2;srf;srw;tif;tiff;x3f;jxl;
```
I think this should be in alphabetical order.

xiota replied:
Is there a tool to keep the values in `ParseExtensionsEnabled` with whichever extensions they belong to?

Lawrence37 replied:
I don't think so. It has to be verified manually.

xiota replied:
I put it in a csv file to look at it in LibreOffice... One of the disabled extensions is png.  Why is that?

Lawrence37 replied:
Not sure. I do remember png as one of the disabled extensions.

xiota replied:
I moved `jxl` after `jpeg`.  I didn't check that the others are alphabetized though.

Thanatomanic commented:
> I think we can enable png per default now. 

Yes, but not in this PR imo (keep it cleanly focused on JXL).

> Someone else will have to make the changes needed to make cmake detect and load libjxl as needed. (Right now it's required.)

@xiota Well, this is an issue indeed. I found out that on Windows MSYS2 does not provide the `libjxl` package yet. So anyone on Windows should first manually compile `libjxl` or we should start working with submodules or something. Not very convenient.
My suggestion is to create a build option `WITH_JXL` here https://github.com/Beep6581/RawTherapee/blob/0e122a485a2339bf2331a7c54a2f5cb2ee2f2dd3/CMakeLists.txt#L197-L225
and set it to OFF by default.
Then you should be able to conditionally check for the package, like here
https://github.com/Beep6581/RawTherapee/blob/0e122a485a2339bf2331a7c54a2f5cb2ee2f2dd3/CMakeLists.txt#L496-L498

But I'm not at all familiar with the intricacies of CMake, so perhaps this doesn't work at all...

EwoutH commented:
Thanks for this effort! I would love to have JPEG XL import support in RawTherapee, what's needed to move this PR forward?

Beep6581 commented:
Builds fine in Manjaro against libjxl-0.7.0 and libjxl_threads-0.7.0, and opens a sample .jxl file fine, though without any metadata.

Thanatomanic commented:
> though without any metadata

Is it to be expected that once we support evix2 this automagically works for jxl too?

kmilos commented:
> Is it to be expected that once we support evix2 this automagically works for jxl too?

If exiv2 is built w/ BMFF support, then yes.

Lawrence37 commented:
Converting this to a draft because the GitHub Actions workflows still need to be updated to use the JXL library. Making JXL optional may also be necessary for the moment.

Beep6581 commented:
Moving to 6.0 due to lack of activity.

EwoutH commented:
@xiota Would you be able to further work on this? If so, is there anything you need to move this forward?

xiota commented:
Issues I think holding this back:

* [ ] Conditional compilation to build libjxl support only when the library is present.
* [ ] Windows library availability?
* [ ] I'd need a refresher on how rawtherapee works
* [ ] I'd need a refresher on how libjxl works
* [ ] update and rebase to current rawtherapee
* [ ] update for current libjxl
* [ ] I don't understand how libjxl color management works

kmilos commented:
> Windows library availability?

No problem w/ this one: https://packages.msys2.org/base/mingw-w64-libjxl

xiota commented:
I force pushed to rebase to the current dev branch.  It's updated to the current libjxl API (only one function changed).  I also figured out how to make libjxl optional, so if it's not installed, compilation should continue with jpeg xl disabled.

kmilos commented on CMakeLists.txt:
```diff
@@ -552,6 +553,16 @@ if(WITH_SYSTEM_KLT)
     find_package(KLT REQUIRED)
 endif()
 
+set(JXL_LIBS "libjxl libjxl_threads")
+if(WITH_JXL)
+    pkg_check_modules(JXL REQUIRED IMPORTED_TARGET ${JXL_LIBS})
+else()
+    pkg_check_modules(JXL IMPORTED_TARGET ${JXL_LIBS})
+endif()
+if(JXL_FOUND)
+    add_definitions(-DLIBJXL)    
+endif()
```
Not a fan of this logic... If the user is explicitly expected to set WITH_JXL to ON, shouldn't silently do stuff behind their back when they no longer don't - they'll still get it built when they change it back to OFF just because the library is present on their system.

In fact, it might be better to default this to ON (you've added extension support), and enable the user to explicitly turn it OFF completely then.

xiota replied:
Changed to tristate with settings: AUTO/ON/OFF

xiota commented:
@Lawrence37 Would you mind running this through workflows to check that the cmake files are compatible?

I updated them to add a tristate option `WITH_JXL`.  Default is AUTO, which enabled JXL import when libjxl is available.  YES force enables the feature, showing error when libjxl is not available.  NO force disables the feature, even when libjxl is found.

I looked at the workflow files, and it seems for Ubuntu and MSYS, libjxl can be added to the depends lists.  However, Ubuntu has libjxl-dev 0.7.0, while MSYS has libjxl 0.10.2.  When I rebased, I updated and test built with 0.10.2.

Lawrence37 commented:
The workflows will probably fail without installing libjxl, but I'll run them anyway.

xiota commented:
@Lawrence37 Thanks for running the workflows earlier.  They all passed (deps not installed and build completed with jxl export disabled).

Would you mind running the workflows again?  I added libjxl to the dep lists.  Hoping they'll build with jxl export enabled.

xiota commented:
Looks like... [macOS](https://github.com/Beep6581/RawTherapee/actions/runs/8668458298?pr=6367) and [Windows](https://github.com/Beep6581/RawTherapee/actions/runs/8668458284?pr=6367) workflows succeeded.  I suppose people using those OSes can grab the artifacts to test?

The Linux workflows failed because they're running on Ubuntu 20.04 and 22.04, but [libjxl](https://packages.ubuntu.com/search?keywords=libjxl-dev&searchon=names&suite=all&section=all) was added in 23.04.  Should be available in 24.04 in a few weeks.

How would RT devs like to proceed?

Lawrence37 commented:
It seems like there is a deb package for Ubuntu 20.04 (https://github.com/libjxl/libjxl/releases/tag/v0.10.2). We can use that or compile libjxl in the AppImage workflow. Note that the runner must be the oldest LTS Ubuntu available (currently 20.04) to maximize compatibility of the AppImage.

xiota commented:
Is figuring out the deps in the workflows important for this PR or could it be deferred?

Lawrence37 commented:
The AppImage workflow generates the official AppImage for releases, so it is important that libjxl is included. Otherwise, there will be no jxl support for AppImage users.

xiota commented:
I found the button to run workflows and got the [AppImage workflow](https://github.com/xiota/RawTherapee/actions/runs/8677760467) to complete successfully.  The resulting appimage was able to thumbnail and load a jxl file.

It uses debs from the libjxl github repo, but libhwy is from a PPA.  The highway repo doesn't publish debs, and the 22.04 deb won't install on 20.04.  The debian highway package has the same problem.

I'm unable to test the codeql workflow.  Assuming no mistakes, it should also work.  It uses libjxl from github, and libhwy from the official Ubuntu 22.04 repo.

Lawrence37 commented:
Nice! Getting a package from a PPA is less than ideal. I see highway can be installed using `vcpkg install highway`. `vcpkg` comes preinstalled in the Ubuntu 20.04 environment. I have never used vcpkg before, but it might be worth a shot.

xiota commented:
Looks like vcpkg might have to be built from source.  Instead of doing that, I decided to try using [`makedeb`](https://www.github.com/makedeb/makedeb) to build, package, and install libjxl.  The resulting appimage seems to work.

kmilos commented:
FWIW, the libjxl project publishes deb packages for 20.04: https://github.com/libjxl/libjxl/releases/tag/v0.10.2

xiota commented:
@kmilos See preceding comments for why they are "less than ideal".

kmilos commented:
I though that was related to PPAs... But I guess that source, although "official", can be considered as one. Then again, I don't see how vcpkg would be different in essence. They are all 3rd parties...

xiota commented:
The libjxl release debs depend on libhwy, which is not in the 20.04 repos.  The only libhwy deb I found is from an old PPA.  The package itself is probably fine because canonical controls the build server.  But if the maintainer changes or deletes the repo, any workflows depending on it could break.

Ubuntu 22.04 does have libhwy in its repos, so the codeql workflow can use the libjxl release debs.

xiota commented:
When I searched earlier for how to use vcpkg, I found instructions about how to build from source or use custom actions that appear to build from source.  However, running a test workflow that calls `vcpkg install libjxl` does appear to be working.  So if preferred, I can try switching the workflow to vcpkg.

xiota commented:
Cannot use vcpkg...
```
-- Checking for modules 'libjxl;libjxl_threads'
--   No package 'libjxl' found
--   No package 'libjxl_threads' found
```

Lawrence37 commented:
To clarify my previous comments:
Using the libjxl debs is fine because they are official. libjxl-dev depends on libhwy and libhwy-dev, which are not available on Ubuntu 20.04 and have no official debs. I saw on the project's GitHub that highway can be installed through vcpkg, and the Ubuntu 20.04 runner has vcpkg preinstalled. My recommendation is to try vcpkg for installing libhwy, after which the libjxl debs can be installed. I have not used vcpkg before, so I don't know if it will work well. I'm also unsure if installing highway through vcpkg will provide all the dependencies needed by libjxl-dev because I don't know if the package manager will see the highway installation. Because of the uncertainty, I'm not demanding that vcpkg be used, but I do think it is worth trying.

xiota commented:
vcpkg is a parallel system to apt, so whatever it installs won't satisfy apt dependencies.  It also installs somewhere that the build scripts don't normally use.  (See error message in [noted in previous comment](https://github.com/Beep6581/RawTherapee/pull/6367#issuecomment-2053992121).)  Since there's already a working solution that doesn't use a PPA, I'm not very interested in figuring out how to make vcpkg work.  Also, 20.04 is EOL next year.  22.04 has libhwy.  24.04 should have libjxl.

Lawrence37 commented:
@xiota I just reviewed the current workflow and I think it is fine because all dependencies either come from the official source or Ubuntu.

xiota commented:
To do?

- [ ] More user testing?
- [ ] Code review?
- [ ] Run `astyle` on added code sections.
- [ ] Anything else?

Lawrence37 commented:
Code review and testing by developers are left. Astyle and static checkers like cppcheck are optional. You can use the official forum at https://discuss.pixls.us/c/software/rawtherapee/14 to generate publicity.

xiota commented:
> You can use the official forum ... to generate publicity.

[RawTherapee JPEG XL Import](https://discuss.pixls.us/t/rawtherapee-jpeg-xl-import/43189)

Floessie commented on rtengine/imageio.cc:
```diff
@@ -24,13 +24,20 @@
 #include <string>
 #include <utility>
 #include <vector>
-
```
Please reapply this blnak line, as it seperates C++ headers from C headers.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -24,13 +24,20 @@
 #include <tiffio.h>
 
+#ifdef LIBJXL
+#include <fstream>
```
Either add a blank line after the C++ header (vs. C headers below) or drop the header in favor of other tools (see my next comment).

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+    format.align = 0;
+
+    // get file contents
+    std::ifstream instream(fname.c_str(), std::ios::in | std::ios::binary);
```
I'm not sure, if this is the right way to deal with the `Glib::ustring` here. Aren't there safer GTK/Glib alternatives to read a file designated by a unicode filename? If so, you can drop the `<fstream>` include above.

xiota replied:
glib has a function to read files, but it's pointer heavy.
```
gboolean
g_file_get_contents (
  const gchar* filename,
  gchar** contents,
  gsize* length,
  GError** error
)
```

Floessie replied:
How about `Gio::File` that is used in various places? But maybe @Lawrence37 has an idea on how problematic the `filename.c_str()` actually is...

xiota replied:
I found this, but I'm pretty sure there's a pointer to remember to free: `bool Gio::File::load_contents(char *&contents, gsize &length);`

xiota replied:
This seems to work... don't know if it's safe passing everything through std::string.
```C++
std::string const contents = Glib::file_get_contents(fname);
std::vector<std::uint8_t> const compressed(contents.begin(), contents.end());
```

Floessie replied:
Hmm, `Glib::file_get_contents(fname)` takes a `std::string` not a `Glib::ustring` and also thows. Additionally, this doubles the amount of memory used. So no, this is not a solution, I'm afraid.

xiota replied:
Would be nice to have a function to read a file into a vector that could be used with all decoders that work with memory buffers.  Would such a function go in rtengine/utils.cc?  Maybe I should look...

I think I misunderstood your initial comment (again).  I didn't notice any issues reading files with Chinese and Japanese filenames using either fstream/c_str or file_get_contents.  I don't recall the reason for using c_str().  I vaguely remember having problems with casting or auto conversion to std::string not working.  Might be a glitch that's been resolved by now.

Lawrence37 replied:
`Glib::ustring::c_str()` returns a UTF-8 encoded Unicode string, but `std::ifstream`'s expected encoding is platform-dependent (I think?). Maybe [`Glib::filename_from_utf8`](https://developer-old.gnome.org/glibmm/2.48/group__CharsetConv.html#ga2bd94064ad97b43324a7854b62f0835b) can be used to ensure the correct encoding for the file name? I'm not sure if the encoding is meant for the platform, Gio, or both. For a full Gio method, you can try getting a file with [`Gio::File::create_for_path`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1File.html#a1b1a4553b226689d50b9525488223e25), opening an input stream with [`Gio::File::read`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1File.html#a0395b3b0b8cec5dc1fd9953e52e5fbc1), and finally read the file content into your buffer with [`Gio::InputStream::read_all`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1InputStream.html#a6da0fc57d1a5df7107370c67e84aa496).

xiota replied:
I added a function to `utils.cc` because it's potentially useful elsewhere.  It uses `Glib::filename_from_utf8` to convert ustring to std::string.  It returns an empty vector if there's any error.

```c++
std::vector<std::uint8_t> getFileData(const Glib::ustring &filename);
```

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+    }
+
+    // grand decode loop...
+    JxlDecoderSetInput(dec.get(), compressed.data(), compressed.size());
```
This takes `const uint8_t* data`, so please make `compressed` const.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+
+    // get file contents
+    std::ifstream instream(fname.c_str(), std::ios::in | std::ios::binary);
+    std::vector<uint8_t> compressed(
```
We use `<cstdint>`, so this should be `std::uint8_t`.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+            // check for ICC profile
+            deleteLoadedProfileData();
+            embProfile = nullptr;
+            size_t icc_size = 0;
```
`std::size_t`.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+                g_printerr("Warning: Empty ICC data.\n");
+            }
+        } else if (status == JXL_DEC_NEED_IMAGE_OUT_BUFFER) {
+            format.data_type = JXL_TYPE_FLOAT;
```
Is this a precaution, or does `format.data_type` really change in the process?

xiota replied:
I don't remember why I did that.  I can change to `assert()`.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+
+    std::vector<uint8_t> icc_profile;
+
+    gpointer buffer = nullptr;
```
We shouldn't use raw pointers anymore. There are two superior solutions, IMHO:

1. `std::vector<unsigned char>`
2. `std::unique_ptr<unsigned char[]>`

xiota replied:
Could you point me to usage examples?  Is it safe to feed vector into a function that takes buffer pointer and size arguments?

Floessie replied:
Yes, it is. And if I'm not mistaken, you already do that with `icc_profile`. :wink:

xiota replied:
My left brain doesn't know what my right brain is doing.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+                return IMIO_READERROR;
+            }
+
+            buffer = g_malloc(buffer_size);
```
No raw pointers, please. See comment above.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+            // may be decoded. This example only keeps the first one.
+        } else if (status == JXL_DEC_SUCCESS) {
+            // All decoding successfully finished.
+            // It's not required to call JxlDecoderReleaseInput(dec.get())
```
This comment seems redundant, as `dec` obviously is a smart pointer.

xiota replied:
The decode loop, including a variation of that comment, is from libjxl examples.

`JXL_DEC_SUCCESS` and "All decoding successfully finished." seems more obvious.

What about shortening it to: "Decoder will be released automatically."

Floessie replied:
That's fine. :+1:

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+            // since the decoder will be destroyed.
+            break;
+        } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+            g_printerr("Error: Already provided all input\n");
```
What about the more common "Premature end of file"?

xiota replied:
The decode loop is from libjxl examples.  Those are the status messages they used. In `jxl/decode.h`, I see nothing corresponding to "premature end of file".  The closest looks like `JXL_DEC_NEED_MORE_INPUT`, but that only occurs when an embedded preview is requested.  Since this isn't asking for a preview, that status would/should never occur.  I guess the result would be a partially decoded image or `JXL_DEC_ERROR`.

Floessie replied:
Okay. If I read "Error: Already provided all input" somewhere this would leave me at a loss. Glad to hear that this should never occur. :sweat_smile:

xiota replied:
Assuming I'm reading the documentation correctly.  I'll try to test with some truncated files to see what happens.

xiota replied:
Maybe, `JXL_DEC_NEED_MORE_INPUT` == "premature end of file"?

(I think I didn't understand the intent of your initial comment.)

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+
+    allocate(width, height);
+
+    int line_length = width * 3 * 4;
```
This should be unsigned, too.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+
+    int line_length = width * 3 * 4;
+
+    for (size_t row = 0; row < height; ++row) {
```
`std::size_t`

Floessie commented on rtengine/imageio.h:
```diff
@@ -90,6 +90,10 @@ class ImageIO : virtual public ImageDatas
     int save (const Glib::ustring &fname) const;
 
+#ifdef LIBJXL
+    int loadJxl (const Glib::ustring &fname);
```
Hmm, the other `load...()` extensions are uppercase. And better append the new one rather than prepend it.

Floessie commented on rtengine/utils.cc:
```diff
@@ -237,6 +237,14 @@ bool hasJpegExtension(const Glib::ustring& filename)
+bool hasJxlExtension(const Glib::ustring& filename)
+{
+   const Glib::ustring extension = getFileExtension(filename);
+   return extension == "jxl";
```
This can be a one-liner: `return getFileExtension(filename) == "jxl";`

Floessie commented on rtgui/thumbnail.cc:
```diff
@@ -205,53 +204,40 @@ void Thumbnail::_generateThumbnailImage ()
-            quick = true;
-            tpp = rtengine::Thumbnail::loadQuickFromRaw (fname, sensorType, tw, th, 1, TRUE);
-        }
+    if ( tpp == nullptr ) {
```
`if (!tpp) {`

xiota replied:
Okay... but... I didn't write that line...
```
$ git blame -L241,+1 0fdac3797aebbba4a751184e60278e5460bc56b7 -- rtgui/thumbnail.cc
0731975ff0 (Flössie 2016-10-12 17:48:40 +0200 241)         if ( tpp == nullptr ) {
```

Floessie replied:
I know, I know. But while you're at it...

And actually, I didn't write that line either, just fired clang-tidy at it. :smiley:

xiota replied:
In this file, there are five instances of `tpp == nullptr` and two of `!tpp`.  Should I change them all to `!tpp`?

(I'm going to change all of them.)

Floessie reviewed (changes requested):
Quick review of the diff with some minor change requests, most prominently the change from raw pointer `buffer` to something more sane. :wink:

xiota commented:
@Floessie Thank you for reviewing.  Let me know if I missed something or there's more.

Floessie commented on rtengine/imageio.cc:
```diff
@@ -967,26 +964,24 @@ int ImageIO::loadJxl(const Glib::ustring &fname)
-    for (size_t row = 0; row < height; ++row) {
-        setScanline (row, ((const unsigned char*)buffer) + (row * line_length), 32);
+    for (std::size_t row = 0; row < height; ++row) {
+        setScanline(row, ((const unsigned char*)buffer.data()) + (row * line_length), 32);
```
I don't think this C-style cast is needed anymore. That's why I suggested `buffer` should be `std::vector<unsigned char>`...

xiota replied:
I will try removing the cast.  If no problems, will bundle with other changes.

Floessie commented:
@xiota It's quite okay from my point of view now. Let @Lawrence37 comment on the `Glib::ustring` problem and functionality.

Thanks for your contribution! :+1:

xiota commented:
Checking in, since it's been nearly a month... anything else needed on my end for this?

Lawrence37 commented:
No, you are just waiting for functionality testing. It's unfortunate that no one has commented about their experience on the Pixls topic. I will test if no one else does, but I do have a backlog of other pull requests to review.

xiota commented:
@Lawrence37 Thanks for letting me know.  No hurry...  Just wanted to know the status.

Lawrence37 commented on rtengine/imageio.cc:
```diff
@@ -822,6 +827,172 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+    std::vector<std::uint8_t> const compressed = getFileData(fname);
+
+    if (compressed.empty()) {
+        g_printerr("Error: loadJXL failed to get data from file\n");
```
Elsewhere we are using `std::cerr` or `fprintf`

Lawrence37 commented:
I got this error compiling with libjxl 0.8.2.
```
/RawTherapee/rtengine/imageio.cc: In member function ‘int rtengine::ImageIO::loadJXL(const Glib::ustring&)’:
/RawTherapee/rtengine/imageio.cc:831:19: error: cannot convert ‘JxlColorProfileTarget’ to ‘const JxlPixelFormat*’
  831 | #define _PROFILE_ JXL_COLOR_PROFILE_TARGET_ORIGINAL
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   JxlColorProfileTarget
/RawTherapee/rtengine/imageio.cc:904:60: note: in expansion of macro ‘_PROFILE_’
  904 |                     JxlDecoderGetICCProfileSize(dec.get(), _PROFILE_, &icc_size)
      |                                                            ^~~~~~~~~
In file included from /usr/include/jxl/decode_cxx.h:20,
                 from /RawTherapee/rtengine/imageio.cc:29:
/usr/include/jxl/decode.h:779:50: note:   initializing argument 2 of ‘JxlDecoderStatus JxlDecoderGetICCProfileSize(const JxlDecoder*, const JxlPixelFormat*, JxlColorProfileTarget, size_t*)’
  779 |     const JxlDecoder* dec, const JxlPixelFormat* unused_format,
      |                            ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/RawTherapee/rtengine/imageio.cc:831:19: error: cannot convert ‘JxlColorProfileTarget’ to ‘const JxlPixelFormat*’
  831 | #define _PROFILE_ JXL_COLOR_PROFILE_TARGET_ORIGINAL
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   JxlColorProfileTarget
/RawTherapee/rtengine/imageio.cc:920:40: note: in expansion of macro ‘_PROFILE_’
  920 |                             dec.get(), _PROFILE_,
      |                                        ^~~~~~~~~
/usr/include/jxl/decode.h:799:50: note:   initializing argument 2 of ‘JxlDecoderStatus JxlDecoderGetColorAsICCProfile(const JxlDecoder*, const JxlPixelFormat*, JxlColorProfileTarget, uint8_t*, size_t)’
  799 |     const JxlDecoder* dec, const JxlPixelFormat* unused_format,
      |                            ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
```

xiota commented:
I test built only with 0.7 and 0.10 because those are the versions readily available in repos.  Looks like the cutoff version I got from examples isn't accurate.
```
#if JPEGXL_NUMERIC_VERSION < JPEGXL_COMPUTE_NUMERIC_VERSION(0, 8, 0)
```

I'll look for the commit that made the change

xiota commented:
Maybe I misinterpreted the something before.  The changes were made in v0.9-snapshot-409-gb08a7049.  darktable uses 0,9,0.  Will change it.

Lawrence37 commented:
They make a note of API changes in their change log: https://github.com/libjxl/libjxl/blob/main/CHANGELOG.md
This can be helpful for determining if there are other places we can improve compatibility. It can also be used for setting a minimum required version if necessary.

Lawrence37 reviewed (commented):
The functionality seem fine. It properly loads the embedded color profiles. EXIF data reads without issue (images without a container load without bugs). HDR images retain data. Animated and transparent images load (with necessary limitations).

2 things that could be improved but do not block this pull request:
- Transparent parts of images are not completely black. There is some weird smearing of colors from opaque areas. See the Dice image from https://jpegxl.info/test-page/.
- JXL status/library version could be added to AboutThisBuild.txt.

xiota commented:
Switched from `g_printerr` to `std::cerr`.  Added libjxl version to `AboutThisBuild.txt`.

I suspect the color smearing may be how the file is actually encoded.  To get rid of it, might have to decode and process the alpha channel.  darktable shows the same effect.  Does RT have any existing code that deals with transparency that I could look at?

Lawrence37 commented:
I don't know if the code is in RawTherapee itself, but transparency is handled for pngs and tiffs.

xiota commented:
> Do you know what it shows in AboutThisBuild.txt if jxl support is disabled or if libjxl is not installed?

Something like: `libjxl: V` (the version number is missing).  Should be the same for the other libraries.  I'll see if it can show something nicer, like `n/a` without being too complicated.

> Users who are upgrading will not see jxl images because their settings carry over.

How is this handled when new raw formats are added?

Lawrence37 commented:
> How is this handled when new raw formats are added?

I don't think this was ever handled. The last addition was .ori, which is disabled by default anyway. The other extension in the git history is .cr3. Nothing special was done for that even though I think something should have been.

xiota commented:
Maybe enabling new file extensions could go in a separate PR that takes care of all of them together?

Handling the transparency case could also go in a separate PR someday, like after jxl export is added.

Lawrence37 commented:
Yes, it's fine to do that in a separate PR.

Lawrence37 commented on CMakeLists.txt:
```diff
@@ -552,6 +555,30 @@ if(WITH_SYSTEM_KLT)
+        set(JXLTHREADS_LIBRARIES "")
+    endif()
+else()
+    set(JXL_VERSION "Not installed")
```
It is not installed at compile time, but libjxl could be installed later or RawTherapee could be distributed to someone who does have libjxl installed. "Disabled" suffices for the user.

xiota replied:
I'd rather keep the strings different.  AboutThisBuild reflects the state of the system at build.  Having different strings is useful to distinguish which branch was taken.  The steps to enable JXL would be slightly different for each.

In a few years, when libjxl is standard across all distros, the AUTO setting can be removed.

Lawrence37 replied:
There is a potential problem with a simple "Not installed" message. Let's say a package distributor shares a build of RawTherapee without jxl support due to libjxl not being installed on the machine used for compiling. A user who wants to open jxl images will find that they don't show up in the file browser. Naturally, they may ask about it (on Pixls, here, or elsewhere) and share their operating system, hardware specifications, where they got RawTherapee from, and the build information  in the splash dialog. They, or someone else, will notice that it says libjxl is not installed. Next they install libjxl and restart their computer, but it still says libjxl is not installed!

This is confusing for people, so it should either be specific that it was not installed during the build process and therefore there is no jxl support, or it should not reference the installation status.

Apologies for the long post, but I hope I'm clear about what my concern is.

xiota replied:
I'll change it to "Disabled (Auto)" for the unavailable / not installed case.  That should prevent the confusion you describe as well as differentiate for later troubleshooting.

xiota replied:
```

# installed, enabled or auto
libjxl: 0.10.2

# installed, disabled (WITH_JXL=FALSE)
libjxl: 0.10.2 - Disabled

# not installed, auto
libjxl: Disabled, Auto

# not installed, disabled (WITH_JXL=FALSE)
libjxl: Disabled
```

Lawrence37 commented:
Looks good to me. You are welcomed to add yourself to AUTHORS.txt.

xiota commented:
I'll leave it alone for now... can always add in a future...

Does github have a way to track todos? jxl export, enable new file types in config, process jxl alpha channel

Lawrence37 commented:
> Does github have a way to track todos?

You can just create issue(s) for them. If you want just one issue, GitHub will pick up any checklist in the original post and show a progress widget for it. See #6563 for an example.

Lawrence37 commented:
@Floessie There are a few minor changes since your last review. Do you want some time to go over the recent updates, or are you ok for merging?

Lawrence37 commented:
Merged. Thank you @xiota for working on this!
